src/script/ml.ts

```diff
@@ -105,20 +107,31 @@ export async function trainModel(): Promise<tf.LayersModel | void> {
   const tensorFeatures = tf.tensor(features);
   const tensorLabels = tf.tensor(labels);
   const nn: LayersModel = createModel();
-  const totalNumEpochs = get(settings).numEpochs;
+  const totalNumEpochs = 160;
 
   try {
     await nn.fit(tensorFeatures, tensorLabels, {
       epochs: totalNumEpochs,
       batchSize: 16,
       validationSplit: 0.1,
-      callbacks: {
-        onTrainEnd,
-        onEpochEnd: (epoch: number) => {
-          // Epochs indexed at 0
-          updateTrainingProgress(epoch / (totalNumEpochs - 1));
-        },
-      },
+      callbacks: [
+        new tf.CustomCallback({ onTrainEnd }),
+        new tf.CustomCallback({
+          onEpochEnd(epoch, logs) {
+            // Epochs indexed at 0
+            updateTrainingProgress(epoch / (totalNumEpochs - 1));
+            if (logs) {
+              if (prevLoss && logs.loss && logs.accuracy) {
+                if (logs.loss >= prevLoss && logs.accuracy === 1) {
```

Comment:
Yep -> [055f312](https://github.com/microbit-foundation/ml-trainer/pull/270/commits/055f31206e45b31967fad79de3ed1c8a5884c15a).

I think in practice this is rarely triggered. We can overshoot the best fit, and once we realise we've done this, we can't go back to the best fit. What we could do with is the [built-in earlyStopping callback](https://js.tensorflow.org/api/latest/#callbacks.earlyStopping) to support `restoreBestWeights`, but it's not currently supported.

Reply:
Oh interesting, thanks for the info and link! Sad that it's not currently supported

I noticed that the built-in earlyStopping callback seems to monitor `val_acc` instead of `acc`. It makes me wonder whether we should be using `val_loss` and `val_acc` instead? 🤔

Reply:
It's `val_acc` by default, but you can change the metric to track. My understanding is that you should use `val_loss` and `val_acc` for regression problems, and `loss` and `acc` for classification problems. The [upstream version is now also tracking loss](https://github.com/microbit-foundation/cctd-ml-machine/blob/a90b7f1f1351d42f8958fc4d932b4201f283ef8a/src/script/mlmodels/LayersModelTrainer.ts#L76) to plot it.

Reply:
Ok, that makes sense, thanks!